File: Cpp/LinkedList/DoublyLinkedList/doublyLinkedList.cpp
#include <iostream>

using namespace std;

struct node {
    int data;
    struct node* next;
    struct node* prev;
} *head;

node* createNode(int val){
	struct node *node, *n;

	node = new(struct node);

	node->data = val;
	node->next = NULL;
	node->prev = NULL;

	head = node;

	return node;
}

void insertBegin(){
	int val;
	cout << "Value: ";
	cin >> val;

	if (head == NULL){
		createNode(val);
		return;
	}

	struct node *t;
	t = new(struct node);
	t->data = val;
	t->prev = NULL;
	t->next = head;

	head->prev = t;
	head = t;
}

void insertLast()
{
    struct node *n, *h;
    int val;
    cout << "value: ";
    cin >> val;

	if (head == NULL){
		n = createNode(val);
		return;
	}else {
		n = new(struct node);
		h = head;
  	 	while (h->next != NULL)
  	  	{
 	       h = h->next;
	    }

  		h->next = n;
   		n->next = NULL;
		n->data = val;
		n->prev = h;
	}    
}

void insertInPosition(){
	struct node  *n, *p;

	int val, pos;
	cout << "Value: ";
    cin >> val;
	cout << "Position: ";
    cin >> pos;

	if (head == NULL)
		return;
	
	p = head;
	for (int i=0; i<pos-1; i++){
		p = p->next;
		if(p == NULL){
			cout << "Sorry, the position you entered is wrong.";
		}
	}

	n = new(struct node);
	n->data = val;
	if(p->next == NULL){
		p->next = n;
		n->next = NULL;
		n->prev = p;
	}else{
		n->next = p->next;
		p->next->prev = n;
		p->next = n;
		n->prev = p;
	}
}

void deleteInPosition()
{
    struct node *s, *prev;
    int pos, i, count = 0;

    if (head == NULL)
        return;

    cout << "position: ";
    cin >> pos;

    s = head;
    if (pos == 1)
        head = s->next;
    else
    {
        while (s != NULL)
        {
            s = s->next;
            count++;
        }

        if (pos >= 2 && pos <= count)
        {
            s = head;
            for (i = 1; i <= pos - 1; i++)
            {
                prev = s;
                s = s->next;
            }
            prev->next = s->next;
        }
        else
            cout << "out of range" << endl;

        free(s);
    }
}

void display()
{
    struct node *s;
    if (head == NULL)
    {
        cout << "Empty" << endl;
        return;
    }

    s = head;
    cout << "Elements: " << endl;
    while (s != NULL)
    {
        cout << s->data << "<->";
        s = s->next;
    }
    cout << "NULL" << endl;
}

void listMenu(){
    cout << "1. Insert in begin" << endl;
    cout << "2. Insert in last" << endl;
    cout << "3. Insert in position" << endl;
    cout << "4. Delete in position" << endl;
    cout << "5. Dispaly linkedlist" << endl;
    cout << "6. Exit" << endl;
    cout << "\n Enter a number between 1 and 6: " << endl;
}

int main(){
	int num;

    // Start with empty list
    head = NULL;

    while (1)
    {
        listMenu();
        cin >> num;

        switch (num)
        {
        case 1:
            insertBegin();
            cout << endl;
            break;
        case 2:
            insertLast();
            cout << endl;
            break;
        case 3:
            insertInPosition();
            cout << endl;
            break;
        case 4:
            deleteInPosition();
            cout << endl;
            break;
        case 5:
            display();
            cout << endl;
            break;
        case 6:
            exit(1);
            cout << endl;
            break;    

        default:
            cout << "Please enter a valid number between 1 to 6";
            break;
        }
    }

    return 0;

}

File: Go/LinkedList/DoublyLinkedList/doublyLinkedList.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Node struct {
	data int
	next *Node
	prev *Node
}

var head *Node

func createNode(val int) *Node {
	var node *Node

	node = new(Node)

	node.data = val
	node.next = nil
	node.prev = nil

	head = node

	return node
}

func insertBegin() {
	var val int
	fmt.Print("Value: ")
	fmt.Scan(&val)

	if head == nil {
		createNode(val)
		return
	}

	var node *Node
	node = new(Node)
	node.data = val
	node.prev = nil
	node.next = head

	head.prev = node
	head = node
}

func insertLast() {
	var (
		n, h *Node
		val  int
	)

	fmt.Print("value: ")
	fmt.Scan(&val)

	if head == nil {
		n = createNode(val)
		return
	} else {
		n = new(Node)
		h = head

		for h.next != nil {
			h = h.next
		}

		h.next = n
		n.next = nil
		n.data = val
		n.prev = h
	}

}

func insertInPosition() {
	var (
		n, p     *Node
		val, pos int
	)

	fmt.Print("Value: ")
	fmt.Scan(&val)
	fmt.Print("Position: ")
	fmt.Scan(&pos)

	if head == nil {
		return
	}

	p = head
	for i := 0; i < pos-1; i++ {
		p = p.next
		if p == nil {
			fmt.Println("Sorry, the position you entered is wrong.")
		}
	}

	n = new(Node)
	n.data = val
	if p.next == nil {
		p.next = n
		n.next = nil
		n.prev = p
	} else {
		n.next = p.next
		p.next.prev = n
		p.next = n
		n.prev = p
	}
}

func display(node *Node) {
	s := node

	fmt.Println("Elements: ")
	for {
		if s == nil {
			break
		}
		fmt.Printf("%d<->", s.data)
		s = s.next
	}
	fmt.Println("NULL")
}

func listMenu() {
	fmt.Println("1. Insert in begin")
	fmt.Println("2. Insert in last")
	fmt.Println("3. Insert in position")
	fmt.Println("4. Delete in position")
	fmt.Println("5. Display linkedlist")
	fmt.Println("6. Exit")
	fmt.Println("Enter a number between 1 and 6:")
}

func main() {
	var num int
	listMenu()

	for {
		fmt.Scan(&num)
		if num == 0 {
			fmt.Println("Thank You, exiting.....")
			break
		} else {
			switch num {
			case 1:
				insertBegin()
				fmt.Println("press 'Enter' to back to menu....")
				bufio.NewReader(os.Stdin).ReadBytes('\n')
				listMenu()
			case 2:
				insertLast()
				fmt.Println("press 'Enter' to back to menu....")
				bufio.NewReader(os.Stdin).ReadBytes('\n')
				listMenu()
			case 3:
				insertInPosition()
				fmt.Println("press 'Enter' to back to menu....")
				bufio.NewReader(os.Stdin).ReadBytes('\n')
				listMenu()
			case 4:
				// deleteInPosition()
				fmt.Println("press 'Enter' to back to menu....")
				bufio.NewReader(os.Stdin).ReadBytes('\n')
				listMenu()
			case 5:
				display(head)
				fmt.Println("press 'Enter' to back to menu....")
				bufio.NewReader(os.Stdin).ReadBytes('\n')
				listMenu()
			case 6:
				os.Exit(0)
			default:
				fmt.Println("Please enter a valid number between 1 to 6")
			}
		}
	}
}
